### PR TITLE
IBP-2792 IBP-2958 IBP-1193 roles permissions

### PR DESCRIPTION
**Pull request by clarysabel**
https://github.com/IntegratedBreedingPlatform/BreedingManager/pull/313
https://github.com/IntegratedBreedingPlatform/Fieldbook/pull/808
https://github.com/IntegratedBreedingPlatform/Commons/pull/227
https://github.com/IntegratedBreedingPlatform/Workbench/pull/280
https://github.com/IntegratedBreedingPlatform/BMSAPI/pull/222
https://github.com/IntegratedBreedingPlatform/Middleware/pull/652

### CODE REVIEW

File: src/main/java/org/ibp/api/domain/role/PermissionDto.java [hunk 1]
```diff
@@ -11,6 +12,12 @@
 
 	private String description;
 
+	private String name;
+
+	private PermissionDto parent;
+
+	private WorkbenchSidebarCategoryLink sidebarCategoryLink;
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/domain/role/PermissionDto.java]: @vmaletta avoid using entities ad part of a DTO

File: src/main/java/org/ibp/api/rest/dataset/DatasetResource.java [hunk 2]
```diff
@@ -44,7 +41,8 @@
 @Api(value = "Dataset Services")
 @Controller
 @RequestMapping("/crops")
-public class DatasetResource {
+public class
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/rest/dataset/DatasetResource.java]: @vmaletta please revert this change

File: src/main/java/org/ibp/api/rest/role/RoleGeneratorInput.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+	private String name;
+
+	private String description;
+
+	private Integer roleType;
+
+	private List<Integer> permissions;
+
+	private boolean active;
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/rest/role/RoleGeneratorInput.java]: @vmaletta should we create it active = true by default?

File: src/main/java/org/ibp/api/rest/role/RoleResource.java [hunk 4]
```diff
@@ -30,4 +31,14 @@
 		return new ResponseEntity<>(this.roleService.getRoles(searchDTO), HttpStatus.OK);
 	}
 
+	@ApiOperation(value = "Save role", notes = "Save role. ")
+	@RequestMapping(value = "/roles", method = RequestMethod.POST)
+	@ResponseBody
+	public ResponseEntity createRole(@RequestBody final RoleGeneratorInput dto) {
+		Preconditions.checkNotNull(dto.getName(), "name cannot be null");
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/rest/role/RoleResource.java]: @vmaletta Please do not use preconditions to validate the input. Move both to the role validator function

File: src/main/java/org/ibp/api/rest/role/RoleValidator.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+		if (roleGeneratorInput.getName().length() > 100) {
+			errors.reject("role.name.lenght");
+		}
+
+		if (this.workbenchDataManager.getRoleByName(roleGeneratorInput.getName()) != null) {
+			errors.reject("role.name.already.exists");
+		}
+
+		if (this.workbenchDataManager.getRoleByDescription(roleGeneratorInput.getDescription()) != null) {
+			errors.reject("role.description.already.exists");
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/rest/role/RoleValidator.java]: @vmaletta is part of the AC definition that description must be unique?

File: src/main/java/org/ibp/api/rest/role/RoleValidator.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+		}
+
+		if (!roleGeneratorInput.getDescription().isEmpty()) {
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/rest/role/RoleValidator.java]: @vmaletta isEmpty check is not needed here, maybe that description is != null

File: src/main/resources/messages_en.properties [hunk 7]
```diff
@@ -252,4 +252,13 @@ user.invalid.instance.role=Role {0} is an instance role and has either crop or p
 user.invalid.crop.role=Role {0} is a crop role and either does not have a crop or has a program selected. Please check
 user.invalid.program.role=Role {0} is a program role and does not have either crop or program. Please check
 user.invalid.crop.program.pair=Program {0} does not exist for crop: {1} 
-user.can.have.one.program.role.per.program=User can only have one program role per program
\ No newline at end of file
+user.can.have.one.program.role.per.program=User can only have one program role per program
+
+role.name.lenght=Role's name exceeds 100 characters
+role.description.lenght=Role's description exceeds 255 characters
+role.roletype.null=Role type could not be empty
```
Comment (clarysabel) [inline src/main/resources/messages_en.properties]: @vmaletta use role.type instead of roletype

File: src/main/resources/messages_en.properties [hunk 8]
```diff
@@ -252,4 +252,13 @@ user.invalid.instance.role=Role {0} is an instance role and has either crop or p
...
+role.roletype.does.not.exist=Role type does not exist
```
Comment (clarysabel) [inline src/main/resources/messages_en.properties]: @vmaletta use role.type instead of roletype

File: src/main/java/org/ibp/api/rest/role/RoleValidator.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+			errors.reject("role.roletype.does.not.exist");
+		}
+		else {
+			for (final Integer permissionId : roleGeneratorInput.getPermissions()) {
+				final Permission permission = this.workbenchDataManager.getPermission(permissionId);
+				if (permission == null) {
+					errors.reject("role.permission.does.not.exist");
+				}
+
+				final Collection result = CollectionUtils.collect(permission.getRoleTypePermissions(), new Transformer() {
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/rest/role/RoleValidator.java]: @vmaletta does it need to be done in a for? Can be done once?

File: src/main/java/org/ibp/api/rest/role/RoleGeneratorInput.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
+package org.ibp.api.rest.role;
+
+import org.pojomatic.annotations.AutoProperty;
+
+import java.util.List;
+
+@AutoProperty
+public class RoleGeneratorInput {
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/rest/role/RoleGeneratorInput.java]: @vmaletta implement toString, equals and hashCode

File: src/main/java/org/ibp/api/java/impl/middleware/role/RoleServiceImpl.java [hunk 11]
```diff
@@ -27,4 +42,60 @@
...
+	private RoleType getRoleType(final Integer roleTypeId) {
+		final RoleType roleType =  this.workbenchDataManager.getRoleType(roleTypeId);
+		return roleType;
+	}
+
+	private List<Permission> getPermission(final List<Integer> permissions) {
+
+		final List<Permission> permissionList = new ArrayList<>();
+		for (final Integer permissionId : permissions) {
+			final Permission permission = this.workbenchDataManager.getPermission(permissionId);
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/java/impl/middleware/role/RoleServiceImpl.java]: @vmaletta better to implement getPermissions and pass a list of Integers

File: src/main/java/org/ibp/api/java/impl/middleware/role/RoleServiceImpl.java [hunk 12]
```diff
@@ -27,4 +42,60 @@
...
+			permissionList.add(permission);
+
+		}
+
+		final Collection result = CollectionUtils.collect(permissionList, new Transformer() {
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/java/impl/middleware/role/RoleServiceImpl.java]: @vmaletta is it this list content equals to permissions, given that you have already validated that all the ids exists?

File: src/main/java/org/ibp/api/java/impl/middleware/role/RoleServiceImpl.java [hunk 13]
```diff
@@ -27,4 +42,60 @@
...
+			@Override
+			public Integer transform(final Object input) {
+				final Permission permission = (Permission) input;
+				return Integer.valueOf(permission.getPermissionId());
+			}
+		});
+
+		for (final Iterator<Permission> it = permissionList.iterator(); it.hasNext();) {
+			final Permission permission = it.next();
+			if (result.contains(permission.getParent().getPermissionId())) {
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/java/impl/middleware/role/RoleServiceImpl.java]: @vmaletta parent can be null

File: src/main/java/org/ibp/api/rest/role/RoleValidator.java (hunk 9 above)
Comment (vmaletta) [reply src/main/java/org/ibp/api/rest/role/RoleValidator.java]: How?

File: src/main/java/org/ibp/api/java/impl/middleware/role/RoleServiceImpl.java (hunk 12 above)
Comment (vmaletta) [reply src/main/java/org/ibp/api/java/impl/middleware/role/RoleServiceImpl.java]: No, I didn't

Comment (clarysabel) [reply src/main/java/org/ibp/api/java/impl/middleware/role/RoleServiceImpl.java]: You did it in the validator

File: src/main/java/org/ibp/api/java/impl/middleware/manager/UserValidator.java [hunk 14]
```diff
@@ -151,7 +144,8 @@ private void validateUserUpdate(Errors errors, UserDetailDto user) {
 					this.validatePersonEmailIfExists(errors, user.getEmail());
 				}
 
-				this.validateCrops(errors, user, userUpdate);
+				// May not be needed with new validations for userRoles
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/manager/UserValidator.java]: Remove commented code and remove unused **validateCrops** method

File: src/main/java/org/ibp/api/java/impl/middleware/manager/UserValidator.java [hunk 15]
```diff
@@ -231,22 +219,157 @@ protected void validatePersonEmailIfExists(final Errors errors, final String eMa
...
-			errors.reject(CANNOT_ASSIGN_SUPERADMIN_ROLE);
-		} else {
-	        this.validateFieldLength(errors, role.getDescription(), ROLE, ROLE_STR, 30);
-		} 
+	protected void validateUserRoles(final Errors errors, final UserDetailDto user) {
+
+		final List<UserRoleDto> userRoles = user.getUserRoles();
+
+		if (!userRoles.isEmpty()) {
+
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/manager/UserValidator.java]: This is one very long method. We can extract related validations to separate methods, say validation related to roles, crop, role types, etc

File: src/main/java/org/ibp/api/security/WorkbenchUserDetailsService.java [hunk 16]
```diff
@@ -46,25 +49,23 @@ public UserDetails loadUserByUsername(String username) throws UsernameNotFoundEx
...
-		final List<GrantedAuthority> authorities = new ArrayList<GrantedAuthority>();
-		if (workbenchUser != null) {
-			final List<UserRole> userRoles = workbenchUser.getRoles();
-			if (userRoles != null && !userRoles.isEmpty()) {
-				for (final UserRole role : userRoles) {
-					authorities.add(new SimpleGrantedAuthority(role.getRole().getCapitalizedRole()));
-				}
-			}
-		}
+	private Collection<? extends GrantedAuthority> getAuthorities(final WorkbenchUser workbenchUser) {
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/security/WorkbenchUserDetailsService.java]: Question: Why are we maintaining two `WorkbenchUserDetailsService `implementations, one in Commons and one in BMSAPI? The one in Commons was able to pass crop and program to the permissionService.
